DataStructures/BitList.cs

```diff
@@ -0,0 +1,1750 @@
+        /// </summary>
+        /// <param name="arrayIndex">Start position.</param>
+        public void CopyTo(double[] array, int arrayIndex)
+        {
+            ToDoubleArray().CopyTo(array, arrayIndex);
+        }
+
+        /// <summary>
+        /// Copy this instance to the specified array.
+        /// </summary>
+        /// <param name="arrayIndex">Start position.</param>
+        public void CopyTo(decimal[] array, int arrayIndex)
+        {
+            ToDecimalArray().CopyTo(array, arrayIndex);
+        }
+
+        /// <summary>
+        /// Copy this instance to the specified array.
+        /// </summary>
+        /// <param name="arrayIndex">Start position.</param>
+        public void CopyTo(char[] array, int arrayIndex)
+        {
+            ToCharArray().CopyTo(array, arrayIndex);
+        }
+
+        /// <summary>
+        /// Get a the value in the specified index.
+        /// </summary>
+        public bool Get(int index)
+        {
+            if (!IsEmpty && index < bits.Length && index >= 0)
+            {
+                return bits[index];
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Get a range of value in the specified index.
+        /// </summary>
+        public BitList Get(int index, int count)
+        {
+            if (!IsEmpty && index < bits.Length && index >= 0 && index + count <= bits.Length - index && count > 0)
+            {
+                var bools = new List<bool>();
+                for (var i = 0; i < count; i++)
+                {
+                    bools.Add(bits[index + i]);
+                }
+
+                return new BitList(bools.ToArray());
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Set the value in the specified index with a new value.
+        /// </summary>
+        public void Set(int index, bool value)
+        {
+            if (!IsEmpty && index < bits.Length && index >= 0)
+            {
+                bits[index] = value;
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        /// <summary>
+        /// Set the value in the specified index with a new collection of values.
+        /// </summary>
+        public void Set(int index, BitList value)
+        {
+            if (!IsEmpty && index < bits.Length && index >= 0 && index + value.Count <= bits.Length - index && value.Count > 0)
+            {
+                var bools = ToBooleanArray().ToList();
+                bools.RemoveRange(index, value.Count);
+                bools.InsertRange(index, value.ToBooleanArray());
+                bits = new BitArray(bools.ToArray());
+            }
+            else
+            {
+                throw new IndexOutOfRangeException();
```

Comment:
Codacy found an issue: ['System.IndexOutOfRangeException' should not be thrown by user code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5559704)